### DISCUSSION TITLE
What map def attributes must match to update an inner map value?

### DISCUSSION THREAD

**Discussion by dthaler** [microsoft/ebpf-for-windows#460] [Q&A] [answered]
bpf_map_update_elem can be used to store an inner map in a map of type BPF_MAP_TYPE_ARRAY_OF_MAPS or BPF_MAP_TYPE_HASH_OF_MAPS.

Certainly the inner map's type must match that of the template (e.g., via inner_map_idx on older versions of Linux), but what other attributes also must match?
key_size? value_size? max_elements?

**Comment by dthaler** [comment]
@teknoraver @qmonnet  Can you help us answer this?

**Comment by teknoraver** [comment, marked as answer]
Hi all,

when inserting a map into a map of maps, the newly inserted map is checked for equality against the first one inserted:
```c
bool bpf_map_meta_equal(const struct bpf_map *meta0,
                       const struct bpf_map *meta1)
{
        /* No need to compare ops because it is covered by map_type */
        return meta0->map_type == meta1->map_type &&
               meta0->key_size == meta1->key_size &&
               meta0->value_size == meta1->value_size &&
               meta0->map_flags == meta1->map_flags;
}
```
The call chain is:
```
__sys_bpf()
map_update_elem()
bpf_map_update_value()
bpf_fd_array_map_update_elem()
bpf_map_fd_get_ptr()
array_map_meta_equal()
bpf_map_meta_equal()
```

**Reply by dthaler** [reply]
@teknoraver also reported via email that max_entries also must match
